### PR TITLE
Bottling Machine Overhaul

### LINKED ISSUE #5294: [Feature Request] Allow Bottling Machine to wait until container is filled
### Description

I was told to post this as an issue on the discord, so here it is.

As far as I know, there is no good way to automate jerrycan filling with IE's Bottling Machine. Because the internal fluid buffer of the Bottling Machine is smaller than the capacity of the jerrycan, putting jerrycans on a Bottling Machine line will just give you a bunch of partially filled jerrycans.

There is probably some way for the user to resolve this with complicated redstone mechanisms, but this is unsatisfying and it feels like there should be a better way to automate it using only IE machines.

### Suggestion

Give Bottling Machines a new functionality that can be toggled with the screwdriver. Instead of filling the container with whatever's in the buffer and then moving to the next one, it should wait until the container is fully filled before outputting it.

### Alternatives

* Allow the Bottling Machine to be upgraded with a larger tank or more pumps for more throughput.
* Add a new Hose Pump machine specifically for filling jerrycans (lol)

### PR DESCRIPTION
**Pull request by BluSunrize**
They now use proper multiblock processes and allow for multiple inputs.

I also added the feature to hold containers until full, which closes #5294

Recipe for empty shells was adjusted to use these new capabilities, and I'm still working on some touchups of the manual.

### CODE REVIEW

File: src/main/java/blusunrize/immersiveengineering/common/blocks/metal/BottlingMachineBlockEntity.java [hunk 1]
```diff
@@ -219,33 +178,51 @@ public void onEntityCollision(Level world, Entity entity)
 			BottlingMachineBlockEntity master = master();
 			if(master==null)
 				return;
-			ItemStack stack = itemEntity.getItem();
-			if(stack.isEmpty())
+
+			List<Tuple<ItemEntity, ItemStack>> itemsOnConveyor = level.getEntitiesOfClass(
+					ItemEntity.class, AABB.unitCubeFromLowerCorner(Vec3.atLowerCornerOf(getBlockPos()))
+			).stream().map(itemEntity1 -> new Tuple<>(itemEntity1, itemEntity1.getItem())).toList();
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/blocks/metal/BottlingMachineBlockEntity.java]: I'd replace Tuple either by `Pair` (the one from DFU, everything else in IE uses that) or a quick `record` with proper member names.

File: src/main/java/blusunrize/immersiveengineering/common/blocks/metal/BottlingMachineBlockEntity.java [hunk 2]
```diff
@@ -404,168 +392,144 @@ public Direction[] sigOutputDirections()
...
-							ItemStack ret = FluidUtils.fillFluidContainer(tile.tanks[0], items.get(0), ItemStack.EMPTY, null);
-							if(!ret.isEmpty())
-								items.add(ret);
-						}
-						if(items.size() <= 1)
-							items.add(items.get(0));
-						tile.markContainingBlockForUpdate(null);
-					}
+					master.allowPartialFill = !master.allowPartialFill;
+					ChatUtils.sendServerNoSpamMessages(player, new TranslatableComponent(Lib.CHAT_INFO
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/blocks/metal/BottlingMachineBlockEntity.java]: Status message (whatever they're called, the popup in the middle) instead?

File: src/main/java/blusunrize/immersiveengineering/common/blocks/metal/BottlingMachineBlockEntity.java [hunk 3]
```diff
@@ -404,168 +392,144 @@ public Direction[] sigOutputDirections()
...
+			return true;
 		}
+		return false;
 	}
 
-	public static class BottlingMachineInventoryHandler implements IItemHandlerModifiable
+	public static class MultiblockProcessBottling extends MultiblockProcessInWorld<BottlingMachineRecipe>
 	{
-		BottlingMachineBlockEntity multiblock;
+		private boolean isFilling = false;
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/blocks/metal/BottlingMachineBlockEntity.java]: Move `false`-initialization to other constructor and make `final`?

File: src/main/java/blusunrize/immersiveengineering/common/crafting/serializers/BottlingMachineRecipeSerializer.java [hunk 4]
```diff
@@ -44,10 +44,23 @@ public BottlingMachineRecipe readFromJson(ResourceLocation recipeId, JsonObject
...
 			outputs.add(readOutput(results.get(i)));
 
-		Ingredient input = Ingredient.fromJson(GsonHelper.getAsJsonObject(json, "input"));
+		IngredientWithSize[] ingredients;
+		if(json.has("input"))
+		{
+			ingredients = new IngredientWithSize[]{
+					IngredientWithSize.deserialize(GsonHelper.getAsJsonObject(json, "input"))
+			};
+		}
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/crafting/serializers/BottlingMachineRecipeSerializer.java]: Braces around single-line/statement if

File: src/main/java/blusunrize/immersiveengineering/common/crafting/serializers/BottlingMachineRecipeSerializer.java [hunk 5]
```diff
@@ -56,16 +69,21 @@ public BottlingMachineRecipe readFromJson(ResourceLocation recipeId, JsonObject
 	public BottlingMachineRecipe fromNetwork(ResourceLocation recipeId, FriendlyByteBuf buffer)
 	{
 		List<Lazy<ItemStack>> outputs = PacketUtils.readList(buffer, IERecipeSerializer::readLazyStack);
-		Ingredient input = Ingredient.fromNetwork(buffer);
+		int inputCount = buffer.readInt();
+		IngredientWithSize[] ingredients = new IngredientWithSize[inputCount];
+		for(int i = 0; i < ingredients.length; i++)
+			ingredients[i] = IngredientWithSize.read(buffer);
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/crafting/serializers/BottlingMachineRecipeSerializer.java:73]: `PacketUtils::read/writeList`? I think vanilla has a helper for it now as well.

File: src/main/java/blusunrize/immersiveengineering/common/util/compat/crafttweaker/managers/BottlingMachineRecipeManager.java [hunk 6]
```diff
@@ -64,17 +66,17 @@ public RecipeType<BottlingMachineRecipe> getRecipeType()
 	 * @docParam output <item:minecraft:wooden_pickaxe>
 	 */
 	@ZenCodeType.Method
-	public void addRecipe(String recipePath, IIngredient itemInput, Many<MCTag> fluidTag, IItemStack[] outputs)
+	public void addRecipe(String recipePath, IIngredientWithAmount[] inputs, Many<MCTag> fluidTag, IItemStack[] outputs)
```
Comment (malte0811) [inline src/main/java/blusunrize/immersiveengineering/common/util/compat/crafttweaker/managers/BottlingMachineRecipeManager.java:69]: Does the example script need to be updated as well?

**Review by malte0811 — APPROVED** [review]

File: src/main/java/blusunrize/immersiveengineering/common/crafting/serializers/BottlingMachineRecipeSerializer.java (hunk 5 above)
Comment (BluSunrize) [reply src/main/java/blusunrize/immersiveengineering/common/crafting/serializers/BottlingMachineRecipeSerializer.java:73]: read/writeList would give me a list, I need an array. No point in making a list that then needs to be put an array again.

**Pull request merged**